### ISSUE TITLE
The redis three-master and three-slave clusters often perform active/standby switchover. As a result, the redisson client may report an error (MOVED), which affects services

### ISSUE THREAD

**Issue by killelk** [redisson/redisson#5972]
redis version：6.0.11
redisson version：3.30.0
Due to the poor performance of network or cluster node machines, our redis cluster is often in an unstable state and will have multiple master/standby switches. In this case, there is a certain probability that redisson's cluster topology information will fail to refresh, which is inconsistent with the actual cluster node information, resulting in its request being sent to the standby redis node. redis returns a moved error, but the redisson client still does not refresh its cluster topology. The error cannot be recovered by itself, and the service needs to be restarted to resolve it. The specific error information is as follows:
2024-06-18 05:00:17.538[WARN ][http-nio-8899-exec-11031][][]com.virtual.vdi.server.internal.api.RestGatewayController.queryCvks [99] - [gw-check] query cvk error
org.redisson.client.RedisTimeoutException: Unable to acquire connection! java.util.concurrent.CompletableFuture@2db07dcc[Completed exceptionally]Increase connection pool size. Node source: NodeSource [slot=7685, addr=redis://10.225.91.213:6379, redisClient=null, redirect=MOVED, entry=null], command: null, params: null after 3 retry attempts
	at org.redisson.command.RedisExecutor$1.run(RedisExecutor.java:278)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.run(HashedWheelTimer.java:715)
	at io.netty.util.concurrent.ImmediateExecutor.execute(ImmediateExecutor.java:34)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:703)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:790)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:503)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:750)
2024-06-18 05:00:17.538[WARN ][http-nio-8899-exec-11030][][]com.virtual.vdi.server.internal.api.RestGatewayController.queryCvks [99] - [gw-check] query cvk error
org.redisson.client.RedisTimeoutException: Unable to acquire connection! java.util.concurrent.CompletableFuture@486898fe[Completed exceptionally]Increase connection pool size. Node source: NodeSource [slot=7685, addr=redis://10.225.91.213:6379, redisClient=null, redirect=MOVED, entry=null], command: null, params: null after 3 retry attempts
	at org.redisson.command.RedisExecutor$1.run(RedisExecutor.java:278)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.run(HashedWheelTimer.java:715)
	at io.netty.util.concurrent.ImmediateExecutor.execute(ImmediateExecutor.java:34)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:703)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:790)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:503)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:750)

**Comment by cuijiawei123** [comment]
+1 求处理

**Comment by killelk** [comment]
According to the above description, we attempted to reproduce and found that after such a switch, when using arthas to obtain the vmtool -x 3 --action getInstances --className org.redisson.connection.MasterSlaveEntry object, it contains information about four master nodes. However, in reality, the redis cluster should only have three master nodes. The specific object information is as follows:
[arthas.log](https://github.com/user-attachments/files/16136047/arthas.log)

**Comment by tkatz77** [comment]
I'm trying to build an example that triggers the issue as I find time to see if that can help troubleshoot it.  I have found that it looks like when a replica is promoted to master, redisson logs the change.  However, when I promote the previous master back, there's no logging.  It will then throw errors similar to the above.  I tried this under 3.27.2 and when I promote the previous master back, I see logging showing that it sees the change.

The example I'm working on uses Spring boot and configures the embedded tomcat to use the tomcat session manager (as my production system runs tomcat).  When I promote the replica on the shard that has the session key in it, I see this in the logging:
( :7000 is the initial master, :7001 is initially a replica)

`
2024-07-10T16:10:57.511-04:00  INFO 38740 --- [redissonTest] [isson-netty-1-5] o.redisson.connection.ConnectionsHolder  : 1 connections initialized for 10.1.109.103/10.1.109.103:7001
2024-07-10T16:10:57.967-04:00  INFO 38740 --- [redissonTest] [sson-netty-1-13] o.redisson.connection.ConnectionsHolder  : 24 connections initialized for 10.1.109.103/10.1.109.103:7001
2024-07-10T16:10:57.978-04:00  INFO 38740 --- [redissonTest] [sson-netty-1-13] o.r.cluster.ClusterConnectionManager     : 10.1.109.102/10.1.109.102:7000 master and related slaves: [addr=redis://10.1.109.103:7001] removed
2024-07-10T16:10:57.978-04:00  INFO 38740 --- [redissonTest] [sson-netty-1-13] o.r.cluster.ClusterConnectionManager     : master: redis://10.1.109.103:7001 added for slot ranges: [[0-0], [5462-8192], [13654-16383]]
2024-07-10T16:11:03.046-04:00  INFO 38740 --- [redissonTest] [isson-netty-1-2] o.r.cluster.ClusterConnectionManager     : added slaves detected for master redis://10.1.109.103:7001. current slaves [] last slaves [redis://10.1.109.102:7000] last failed slaves []
2024-07-10T16:11:03.095-04:00  INFO 38740 --- [redissonTest] [sson-netty-1-20] o.redisson.connection.ConnectionsHolder  : 1 connections initialized for 10.1.109.102/10.1.109.102:7000
2024-07-10T16:11:03.565-04:00  INFO 38740 --- [redissonTest] [sson-netty-1-28] o.redisson.connection.ConnectionsHolder  : 24 connections initialized for 10.1.109.102/10.1.109.102:7000
2024-07-10T16:11:03.565-04:00  INFO 38740 --- [redissonTest] [sson-netty-1-28] o.r.cluster.ClusterConnectionManager     : slave: redis://10.1.109.103:7001 added for master redis://10.1.109.102:7000 and slot ranges: [[0-0], [5462-8192], [13654-16383]]
2024-07-10T16:11:03.565-04:00  INFO 38740 --- [redissonTest] [sson-netty-1-28] o.redisson.connection.MasterSlaveEntry   : master 10.1.109.103/10.1.109.103:7001 excluded from slaves
`

When I then promote the previous master back, there's no additional logging showing the change back.  If I then reload the page with the session I'll get this exception:

`
2024-07-10T16:12:36.336-04:00 ERROR 38740 --- [redissonTest] [io-8080-exec-10] o.a.c.c.C.[Tomcat].[localhost]           : Exception Processing [/test]
org.redisson.client.RedisTimeoutException: Unable to acquire connection! java.util.concurrent.CompletableFuture@24a0eda0[Completed exceptionally: java.util.concurrent.CancellationException]Increase connection pool size. Node source: NodeSource [slot=14479, addr=redis://10.1.109.102:7000, redisClient=null, redirect=MOVED, entry=null], command: (HMSET), params: [testing:redisson:tomcat_session:9A2BDB9E0EEE972D000620363FBFEFED, PooledUnsafeDirectByteBuf(ridx: 0, widx: 24, cap: 256), PooledUnsafeDirectByteBuf(ridx: 0, widx: 7, cap: 256), PooledUnsafeDirectByteBuf(ridx: 0, widx: 24, cap: 256), PooledUnsafeDirectByteBuf(ridx: 0, widx: 7, cap: 256)] after 3 retry attempts
	at org.redisson.command.RedisExecutor$1.run(RedisExecutor.java:279) ~[redisson-3.32.0.jar:3.32.0]
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.run(HashedWheelTimer.java:713) ~[netty-common-4.1.111.Final.jar:4.1.111.Final]
	at io.netty.util.concurrent.ImmediateExecutor.execute(ImmediateExecutor.java:34) ~[netty-common-4.1.111.Final.jar:4.1.111.Final]
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:701) ~[netty-common-4.1.111.Final.jar:4.1.111.Final]
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:788) ~[netty-common-4.1.111.Final.jar:4.1.111.Final]
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:501) ~[netty-common-4.1.111.Final.jar:4.1.111.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.111.Final.jar:4.1.111.Final]
	at java.base/java.lang.Thread.run(Thread.java:1583) ~[na:na]
	Suppressed: java.util.concurrent.CancellationException: null
		at org.redisson.command.RedisExecutor$1.run(RedisExecutor.java:275) ~[redisson-3.32.0.jar:3.32.0]
		... 7 common frames omitted
`

This is using Redis 7.0.14, and Redisson 3.32.0

When I perform the same test using Redisson 3.27.2, it outputs logs when I switch the previous master back:

`
2024-07-10T16:21:25.730-04:00  INFO 16040 --- [redissonTest] [sson-netty-1-10] o.redisson.connection.ConnectionsHolder  : 1 connections initialized for 10.1.109.102/10.1.109.102:7000
2024-07-10T16:21:26.260-04:00  INFO 16040 --- [redissonTest] [sson-netty-1-21] o.redisson.connection.ConnectionsHolder  : 24 connections initialized for 10.1.109.102/10.1.109.102:7000
2024-07-10T16:21:26.263-04:00  INFO 16040 --- [redissonTest] [sson-netty-1-21] o.r.cluster.ClusterConnectionManager     : 10.1.109.103/10.1.109.103:7001 master and related slaves: [addr=redis://10.1.109.102:7000] removed
2024-07-10T16:21:26.263-04:00  INFO 16040 --- [redissonTest] [sson-netty-1-21] o.r.cluster.ClusterConnectionManager     : master: redis://10.1.109.102:7000 added for slot ranges: [[0-0], [5462-8192], [13654-16383]]
2024-07-10T16:21:31.422-04:00  INFO 16040 --- [redissonTest] [isson-netty-1-2] o.redisson.connection.ConnectionsHolder  : 1 connections initialized for 10.1.109.103/10.1.109.103:7001
2024-07-10T16:21:31.851-04:00  INFO 16040 --- [redissonTest] [sson-netty-1-10] o.redisson.connection.ConnectionsHolder  : 24 connections initialized for 10.1.109.103/10.1.109.103:7001
2024-07-10T16:21:31.851-04:00  INFO 16040 --- [redissonTest] [sson-netty-1-10] o.r.cluster.ClusterConnectionManager     : slave: redis://10.1.109.103:7001 added for slot ranges: [[0-0], [5462-8192], [13654-16383]]
2024-07-10T16:21:31.851-04:00  INFO 16040 --- [redissonTest] [sson-netty-1-10] o.redisson.connection.MasterSlaveEntry   : master 10.1.109.102/10.1.109.102:7000 excluded from slaves
`

**Comment by tkatz77** [comment]
I created a simple test where rather then using tomcat sessions, I simply set an integer and increment it.  After performing the swap, I'm seeing the same in the logs, where it detects the first failover, but not the second (fail back).  While I do not receive the same exception, further sets do appear to be going to the replica that was previously master, and get the MOVED response.

```
2024-07-11T09:17:49.282-04:00 DEBUG 37128 --- [redissonTest] [io-8080-exec-10] o.r.cluster.ClusterConnectionManager     : slot 14855 for test:counter
2024-07-11T09:17:49.283-04:00 DEBUG 37128 --- [redissonTest] [io-8080-exec-10] org.redisson.command.RedisExecutor       : acquired connection for command (PSETEX) and params [test:counter, 300000, PooledUnsafeDirectByteBuf(ridx: 0, widx: 2, cap: 256)] from slot NodeSource [slot=14855, addr=null, redisClient=null, redirect=null, entry=null] using node 10.1.109.103/10.1.109.103:7001... RedisConnection@282959410 [redisClient=[addr=redis://10.1.109.103:7001], channel=[id: 0xa4d84761, L:/10.10.10.10:9776 - R:10.1.109.103/10.1.109.103:7001], currentCommand=null, usage=1]
2024-07-11T09:17:49.292-04:00 DEBUG 37128 --- [redissonTest] [isson-netty-1-9] org.redisson.command.RedisExecutor       : connection released for command (PSETEX) and params [test:counter, 300000, PooledUnsafeDirectByteBuf(ridx: 0, widx: 2, cap: 256)] from slot NodeSource [slot=14855, addr=null, redisClient=null, redirect=null, entry=null] using connection RedisConnection@282959410 [redisClient=[addr=redis://10.1.109.103:7001], channel=[id: 0xa4d84761, L:/10.10.10.10:9776 - R:10.1.109.103/10.1.109.103:7001], currentCommand=CommandData [promise=java.util.concurrent.CompletableFuture@3b3ec50f[Completed exceptionally: org.redisson.client.RedisMovedException], command=(PSETEX), params=[test:counter, 300000, PooledUnsafeDirectByteBuf(ridx: 0, widx: 2, cap: 256)], codec=org.redisson.codec.Kryo5Codec], usage=0]
2024-07-11T09:17:49.292-04:00 DEBUG 37128 --- [redissonTest] [isson-netty-1-9] org.redisson.command.RedisExecutor       : acquired connection for command (PSETEX) and params [test:counter, 300000, PooledUnsafeDirectByteBuf(ridx: 0, widx: 2, cap: 256)] from slot NodeSource [slot=14855, addr=redis://10.1.109.102:7000, redisClient=null, redirect=MOVED, entry=null] using node 10.1.109.102/10.1.109.102:7000... RedisConnection@600402463 [redisClient=[addr=redis://10.1.109.102:7000], channel=[id: 0x149634b1, L:/10.10.10.10:9810 - R:10.1.109.102/10.1.109.102:7000], currentCommand=null, usage=1]
2024-07-11T09:17:49.301-04:00 DEBUG 37128 --- [redissonTest] [sson-netty-1-32] org.redisson.command.RedisExecutor       : connection released for command (PSETEX) and params [test:counter, 300000, PooledUnsafeDirectByteBuf(ridx: 0, widx: 2, cap: 256)] from slot NodeSource [slot=14855, addr=redis://10.1.109.102:7000, redisClient=null, redirect=MOVED, entry=null] using connection RedisConnection@600402463 [redisClient=[addr=redis://10.1.109.102:7000], channel=[id: 0x149634b1, L:/10.10.10.10:9810 - R:10.1.109.102/10.1.109.102:7000], currentCommand=CommandData [promise=java.util.concurrent.CompletableFuture@5128bbc1[Completed normally], command=(PSETEX), params=[test:counter, 300000, PooledUnsafeDirectByteBuf(ridx: 0, widx: 2, cap: 256)], codec=org.redisson.codec.Kryo5Codec], usage=1]
```

The cluster nodes debug has the correct layout:
```
2024-07-11T09:17:50.646-04:00 DEBUG 37128 --- [redissonTest] [isson-netty-1-2] o.r.cluster.ClusterConnectionManager     : cluster nodes state got from 10.1.109.101/10.1.109.101:7000:
609ce793121b53ed54ef63f741b083f3147e1f49 10.1.109.102:7001@17001 slave c7a9640585e69f9020711e996bb51590409d302b 0 1720703871000 196 connected
158aa7da7b3db2703d3589487fbb9b7464689ee6 10.1.109.103:7000@17000 master - 0 1720703872619 160 connected 8193-13653
c7a9640585e69f9020711e996bb51590409d302b 10.1.109.101:7000@17000 myself,master - 0 1720703872000 196 connected 1-5461
34f19bf915ceb9b5daed2509e057b12ecdbc49cb 10.1.109.101:7001@17001 slave 158aa7da7b3db2703d3589487fbb9b7464689ee6 0 1720703872000 160 connected
129e51c08358e0b781f4179b87ceaf9327e44606 10.1.109.102:7000@17000 master - 0 1720703872519 210 connected 0 5462-8192 13654-16383
e73c4635466e6c204a124676f96ca02b4f877df3 10.1.109.103:7001@17001 slave 129e51c08358e0b781f4179b87ceaf9327e44606 0 1720703872519 210 connected
```

In this case 10.1.109.102:7000 is the master (that was made replica, then master again), but the SET's after the swap back are first going to 10.1.109.103:7001 (that was made master, then back to replica).

This isn't as bad as when it happens in the tomcat session manager (where it ultimately takes down the whole system), but still doesn't seem ideal that it will always try to send to a replica, then have to try again on the master.

**Comment by killelk** [comment]
We tried to downgrade redisson to version 3.27.2 and did not encounter the issue. Also, we would like to inquire about the expected release date for version 3.33.0.

**Comment by killelk** [comment]
Has the cause of this issue been identified?

**Comment by mrniko** [comment]
Yes. I'll make a fix during this month

**Comment by mrniko** [comment]
Fixed.

**Comment by mzygQAQ** [comment]
Which version fix this ? my redisson verson is 2.13.2
When I cyclically switched the master of the Redis cluster, the Redis client seemed to be working normally without any exceptions thrown.
However, in reality, Redis sent a large number of incorrect requests to the wrong nodes, resulting in a large number of Redis moves. At the client level, it seems that the topology will not be updated. The incorrect redirect requests caused the ops encapsulation of the Redis cluster to increase from 100k to nearly 500k/s

**Comment by tkatz77** [comment]
It's fixed in 3.33.0
